presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDataDefinitionExecution.java

```diff
@@ -0,0 +1,110 @@
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.airlift.concurrent.MoreFutures.getFutureValue;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.Objects.requireNonNull;
+
+public class PrestoSparkDataDefinitionExecution<T extends Statement>
+        implements IPrestoSparkQueryExecution
+{
+    private static final Logger log = Logger.get(PrestoSparkDataDefinitionExecution.class);
+
+    private final DataDefinitionTask<T> task;
+    private final QueryStateMachine queryStateMachine;
+    private final T statement;
+    private final TransactionManager transactionManager;
+    private final AccessControl accessControl;
+    private final Metadata metadata;
+
+    public PrestoSparkDataDefinitionExecution(
+            DataDefinitionTask<T> task,
+            QueryStateMachine queryStateMachine,
+            T statement,
+            TransactionManager transactionManager,
+            AccessControl accessControl,
+            Metadata metadata)
+    {
+        this.task = requireNonNull(task, "task is null");
+        this.queryStateMachine = requireNonNull(queryStateMachine, "session is null");
+        this.statement = requireNonNull(statement, "statement is null");
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public List<List<Object>> execute()
+    {
+        log.info("Executing DDL statemt %s", statement.toString());
+
+        ListenableFuture<?> future = task.execute(statement, transactionManager, metadata, accessControl, queryStateMachine, Collections.emptyList());
+
+        Futures.addCallback(future, new FutureCallback<Object>()
+        {
+            @Override
+            public void onSuccess(@Nullable Object result)
+            {
+                getFutureValue(transactionManager.asyncCommit(getTransactionInfo(queryStateMachine.getSession(), transactionManager).getTransactionId()));
+            }
+
+            @Override
+            public void onFailure(Throwable e)
+            {
+                fail(e);
+                throwIfInstanceOf(e, Error.class);
+            }
+        }, directExecutor());
+
+        return Collections.emptyList();
+    }
+
+    private void fail(Throwable cause)
+    {
+        queryStateMachine.transitionToFailed(cause);
+        queryStateMachine.updateQueryInfo(Optional.empty());
```

Comment:
For PoS query state machine should be unnecesary. Have we considered refactoring tasks not to require the state machine?

Reply:
yeah correct  QueryStateMachine is not necessary for PoS. So initially I started refactoring tasks to remove dependency on QueryStateMachine. I was able to successfully refactor Drop,Grant,Rename,Create tasks but the transactional tasks - SetSessionTask,ResetSessionTask,RollbackTask,CommitTask,StartTransactionTask,UseTask,PrepareTask had  dependency on other methods of QuerystateMachine. It involved refactoring all tasks and i was afraid that it may cause regression but we have enough tests to catch these issues. I agree this may be better approach (to reduce dependency on QueryStateMachine). Let me know your thoughts and I can follow that approach.